File: February2022/Python/MajorityElement.py
# TC : O(NlogN)
# Sc : O(1)

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        nums.sort()
        prev = nums[0]
        count = 1
        length = len(nums)
        if length <= 2:
            return nums[0]
        for ind in range(1, length):
            if nums[ind] == prev:
                count += 1
                if count > length//2:
                    return nums[ind]
            else:
                count = 1
                prev = nums[ind]

File: February2022/Java/MajorityElement.java


// @saorav212294

// TC - O(n)
// SC - O(1)

// If an element is majority it will beat out the combined frequency of the rest of the elements in the array


class Solution {
    public int majorityElement(int[] nums) {
        int count = 1, maj = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (maj == nums[i])
                count += 1;
            else {
                count -= 1;
                if (count  < 0) {
                    count *= -1;
                    maj = nums[i];
                }
            }
        }
        return maj;
    }
}
class Solution {

	// TC : O(n)
	// SC :O(1)
	public int majorityElement(int[] nums) {
		int count = 1;
		int majorityEl = nums[0];
		for(int i=1;i<nums.length;i++) {
			int el = nums[i];
			if(el == majorityEl){
				count++;
			} else{
				count--;
			}
			if(count == 0){
				majorityEl = nums[i];
				count =1;
			}
		}

		return majorityEl;

	}
}
